### PR TITLE
[DSM] New payment and punishment menchanism

### LINKED ISSUE #223: [Payment] Returning mechanism
If the `force_transfer` failed(which means the slot pay failed), what should we take?

### LINKED ISSUE #259: [Market] Remove `maybe_punish_merchant` and `close_sorder` interface
Due to the new MPoW, `maybe_punish_merchant` and `close_sorder` is no used.

### PR DESCRIPTION
**Pull request by MyronFanQiu**
Fix #223 
Fix #259

**Mechanism**
- Merchant need to trigger the payment by a extrinsic. Slot payment mechanism has been removed.
- The storage orders can only be closed by `sorder_payment`
- If the current block is lager than the sorder's `expired_on`, `report_works` won't update the sorder's status anymore.

**Code Change**
- Remove `payment` module
- Remove two interfaces `maybe_punishment_merchant` and `close_order` from `MarketInterface`
- Add one interface `update_merchant_punishment` into `MarketInterface`
- Add one more extrinsic `sorder_payment`
- Add some util functions.

**Test**
- Change the expired date for sorders in `swork` module
- Add three more tests for new payment mechanism

### CODE REVIEW

**Comment by zikunfan** [comment]
> Fix #223
> Fix #259
> 
> **Mechanism**
> 
> * Merchant need to trigger the payment by a extrinsic. Slot payment mechanism has been removed.
> * The storage orders can only be closed by `sorder_payment`
> * If the current block is lager than the sorder's `expired_on`, `report_works` won't update the sorder's status anymore.
> 
> **Code Change**
> 
> * Remove `payment` module
> * Remove two interfaces `maybe_punishment_merchant` and `close_order` from `MarketInterface`
> * Add one interface `update_merchant_punishment` into `MarketInterface`
> * Add one more extrinsic `sorder_payment`
> * Add some util functions.
> 
> **Test**
> 
> * Change the expired date for sorders in `swork` module
> * Add three more tests for new payment mechanism

Can we merge `update_merchant_punishment ` and `update_sorder`? Because it just a special status of upcoming sOrder status.

File: cstrml/market/src/lib.rs [hunk 1]
```diff
@@ -559,74 +597,34 @@ impl<T: Trait> Module<T> {
...
-        } 
-        else if punishment.success >= ((*punish_duration as f64) * 0.90) as EraIndex {
-            real_punish_value = (Perbill::from_percent(50)) * so.amount;
-        } 
-        else {
-            real_punish_value = so.amount;
+    pub fn update_merchant_punishment(order_id: &T::Hash, current_block: &BlockNumber, so_status: &OrderStatus) {
+        if let Some(mut p) = Self::merchant_punishments(order_id) {
+            match so_status {
+                OrderStatus::Success => p.success += current_block - p.updated_at,
```
Comment (zikunfan) [inline cstrml/market/src/lib.rs:607]: Is it necessary to record success?

File: cstrml/market/src/lib.rs [hunk 2]
```diff
@@ -773,6 +768,32 @@ impl<T: Trait> Module<T> {
...
+                reward_ratio = Perbill::one();
+            } else if punishment_ratio >= 0.98 {
+                reward_ratio = Perbill::from_percent(95);
+            } else if punishment_ratio >= 0.95 {
+                reward_ratio = Perbill::from_percent(90);
+            } else if punishment_ratio >= 0.90 {
+                reward_ratio = Perbill::from_percent(80);
+            } else if punishment_ratio >= 0.85 {
+                reward_ratio = Perbill::from_percent(50);
+            } else {
```
Comment (zikunfan) [inline cstrml/market/src/lib.rs:791]: if `ratio < 0.85`, sOrder should close immediately, and the remaining amount should turns back to user.

File: cstrml/market/src/lib.rs [hunk 3]
```diff
@@ -559,74 +597,34 @@ impl<T: Trait> Module<T> {
...
+        merchant: &T::AccountId,
+        slash_value: BalanceOf<T>
+    ) {
+        T::Currency::slash(merchant, slash_value);
+        // Update ledger
+        let mut pledge = Self::pledges(merchant);
+        pledge.total -= slash_value;
+        pledge.used -= slash_value;
+        Self::upsert_pledge(merchant, &pledge);
+        //TODO: Move slash value into treasury
```
Comment (zikunfan) [inline cstrml/market/src/lib.rs:627]: Recording to the pm spec,  the specific `TODO` should be `50%` to user, `50%` to treasury

File: cstrml/market/src/lib.rs [hunk 4]
```diff
@@ -549,6 +542,51 @@ decl_module! {
...
+        /// Do storage order payment
+        #[weight = 1_000_000]
+        pub fn sorder_payment(
+            origin,
+            order_ids: Vec<T::Hash>
+        ) {
+            let who = ensure_signed(origin)?;
+
+            // The length of order_ids cannot be too long.
+            ensure!(order_ids.len()<=T::MaxRewardLength::get().try_into().unwrap(), Error::<T>::RewardLengthTooLong);
```
[the remark this answers is not in the document]
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs:555]: Open an issue to track https://github.com/crustio/crust/issues/268

File: cstrml/market/src/lib.rs [hunk 5]
```diff
@@ -549,6 +542,51 @@ decl_module! {
...
+                        if reward_block >= so.expired_on {
+                            Self::close_sorder(&order_id, so.amount - slash_value);
+                        } else {
+                            so.claimed_at = reward_block;
+                            <StorageOrders<T>>::insert(order_id, so);
+                        }
+                    }
+                }
+            }
+            Self::deposit_event(RawEvent::PaymentSuccess(who));
```
[the remark this answers is not in the document]
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs:588]: Rename it to `PaysOderSuccess`

File: cstrml/market/src/lib.rs (hunk 2 above)
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs:791]: Yes. That's the expected behavior of current implementation.

File: cstrml/market/src/lib.rs (hunk 3 above)
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs:627]: We can discuss about this one. According to the doc, it seems like just 50% of pledge currency would be slashed into treasury. The other 50% pledge would be just free and used for incoming storage orders.

File: cstrml/market/src/lib.rs [hunk 6]
```diff
@@ -549,6 +542,51 @@ decl_module! {
...
+
+                    let reward_ratio = Self::get_reward_ratio(&order_id);
+                    let mut slash_value = Zero::zero();
+                    // If it's the badest situation.
+                    if reward_ratio.is_zero() {
+                        slash_value = (Perbill::from_percent(50)) * so.amount;
+                        Self::slash_pledge(&so.merchant, slash_value);
+                        // Set the reward block to the expired on to trigger closing the order
+                        reward_block = so.expired_on;
+                    }
```
[the remark this answers is not in the document]
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs:574]: Nice comments. Merging the logic can provide more extensibility. I think.

File: cstrml/market/src/lib.rs [hunk 7]
```diff
@@ -224,8 +214,8 @@ pub trait Trait: system::Trait {
     /// Minimum storage order duration
     type MinimumSorderDuration: Get<u32>;
 
-    /// Punishment Duration
-    type PunishDuration: Get<EraIndex>;
+    /// Max Reward Length
+    type MaxRewardLength: Get<u32>;
```
[the remark this answers is not in the document]
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs:218]: Rename it to `ClaimLimit`

File: cstrml/market/src/lib.rs [hunk 8]
```diff
@@ -143,10 +144,8 @@ pub trait MarketInterface<AccountId, Hash, Balance> {
     fn maybe_get_sorder(order_id: &Hash) -> Option<StorageOrder<AccountId, Balance>>;
     /// (Maybe) set storage order's status
     fn maybe_set_sorder(order_id: &Hash, so: &StorageOrder<AccountId, Balance>);
-    /// Upsert punish record and (maybe) punish merchant
-    fn maybe_punish_merchant(order_id: &Hash);
-    /// close storage order
-    fn close_sorder(order_id: &Hash);
+    /// Update merchant punishment
+    fn update_merchant_punishment(order_id: &Hash, current_block: &BlockNumber, so_status: &OrderStatus);
```
[the remark this answers is not in the document]
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs]: will merge it

File: cstrml/market/src/lib.rs [hunk 9]
```diff
@@ -549,6 +542,51 @@ decl_module! {
...
+                    let mut slash_value = Zero::zero();
+                    // If it's the badest situation.
+                    if reward_ratio.is_zero() {
+                        slash_value = (Perbill::from_percent(50)) * so.amount;
+                        Self::slash_pledge(&so.merchant, slash_value);
+                        // Set the reward block to the expired on to trigger closing the order
+                        reward_block = so.expired_on;
+                    }
+                    // Do the payment
+                    let reward_amount = Perbill::from_rational_approximation(reward_block - so.claimed_at, so.expired_on - so.completed_on) * so.amount;
```
[the remark this answers is not in the document]
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs]: split this data structure

File: cstrml/market/src/lib.rs (hunk 1 above)
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs:615]: keep it

File: cstrml/market/src/lib.rs [hunk 10]
```diff
@@ -549,84 +546,103 @@ decl_module! {
...
+                        let mut slash_value = Zero::zero();
+                        // If we do need slash the merchant of this sorder.
+                        if !slash_ratio.is_zero() {
+                            slash_value = slash_ratio * so_info.amount;
+                            Self::slash_pledge(&so_info.merchant, slash_value);
+                            // Set the payment block to the expired on to trigger closing the order
+                            payment_block = so_status.expired_on;
+                        }
+                        // Do the payment
+                        let payment_amount = Perbill::from_rational_approximation(payment_block - so_status.claimed_at, so_status.expired_on - so_status.completed_on) * so_info.amount;
```
Comment (zikunfan) [inline cstrml/market/src/lib.rs:589]: `so_status.expired_on - so_status.completed_on` -> `so_info.duration`?

File: cstrml/swork/src/mock.rs [hunk 11]
```diff
@@ -393,23 +380,27 @@ pub fn add_success_sorders(who: &AccountId) {
     ].to_vec();
 
     for (idx, file) in files.iter().enumerate() {
-        insert_sorder(who, file, idx as u8, 350, OrderStatus::Success);
+        insert_sorder(who, file, idx as u8, 1000, OrderStatus::Success);
```
Comment (zikunfan) [inline cstrml/swork/src/mock.rs:383]: Why change this?

File: cstrml/market/src/lib.rs (hunk 10 above)
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs:589]: goot catch. fixed.

File: cstrml/swork/src/mock.rs (hunk 11 above)
Comment (MyronFanQiu) [reply cstrml/swork/src/mock.rs:383]: There is a new mechanism that "If the current block is lager than the sorder's expired_on, report_works won't update the sorder's status anymore.". We can double discuss about this behavior which make senses to me.

Comment (MyronFanQiu) [reply cstrml/swork/src/mock.rs:383]: There is a new mechanism that "If the current block is lager than the sorder's expired_on, report_works won't update the sorder's status anymore.". We can double discuss about this behavior which make senses to me.

**Comment by zikunfan** [comment]
lgtm🥳

**Review by zikunfan — APPROVED** [review]

**Review by LowEntropyBody — APPROVED** [review]

**Pull request merged**